libs/wingsdk/src/cloud/table.ts

```diff
@@ -0,0 +1,134 @@
+import { Construct } from "constructs";
+import { fqnForType } from "../constants";
+import { App, Resource } from "../core";
+import { Json } from "../std";
+
+/**
+ * Global identifier for `Table`.
+ */
+export const TABLE_FQN = fqnForType("cloud.Table");
+
+/**
+ * Table column types
+ */
+export enum ColumnType {
+  /** string type */
+  STRING,
+  /** number type */
+  NUMBER,
+  /** bool type */
+  BOOLEAN,
+  /** date type */
+  DATE,
+  /** json type */
+  JSON,
+}
+
+/**
+ * Properties for `Table`.
+ */
+export interface TableProps {
+  /**
+   * The table's columns.
+   * @default undefined
+   */
+  readonly columns: { [key: string]: ColumnType };
+  /**
+   * The table's primary key. No two rows can have the same value for the
+   * primary key.
+   * @default undefined
+   */
+  readonly primaryKey: string;
+}
+
+/**
+ * Represents a NoSQL database table that can be used to store and query data.
+ * @inflight `@winglang/sdk.cloud.ITableClient`
+ */
+export abstract class Table extends Resource {
+  /**
+   * Create a new `Table` instance.
+   * @internal
+   */
+  public static _newTable(
+    scope: Construct,
+    id: string,
+    props: TableProps
+  ): Table {
+    return App.of(scope).newAbstract(TABLE_FQN, scope, id, props);
+  }
+
+  /**
+   * Table primary key name
+   */
+  public readonly primaryKey: string;
+  /**
+   * Table columns
+   */
+  public readonly columns: { [key: string]: ColumnType };
+  public readonly stateful = true;
+  constructor(scope: Construct, id: string, props: TableProps) {
+    super(scope, id);
+
+    this.display.title = "Table";
+    this.display.description =
+      "A cloud NoSQL database table that can be used to store and query data";
+
+    this.primaryKey = props.primaryKey;
+    this.columns = props.columns;
+  }
+}
+
+/**
+ * Inflight interface for `Table`.
+ */
+export interface ITableClient {
+  /**
+   * Insert a row into the table.
+   * @param row data to be inserted.
+   * @inflight
+   */
+  insert(row: Json): void;
+  /**
+   * Update a row in the table.
+   * @param row data to be updated.
+   * @inflight
+   */
+  update(row: Json): void;
+  /**
+   * Delete a row from the table.
+   * @param key primary key to delete the row.
+   * @inflight
+   */
+  delete(key: any): void;
```

Comment:
I had some doubts about this, will the `primary key` always be a string?

DynamoDB accepts string, number or binary types for primary key

Reply:
I would stay with strings for now (we don't want to over-design to the specifics of DynamoDB or a single cloud provider)

Reply:
I feel like dynamically adding new fields to the table after you create it might be an extra feature that we don't need to support in the API right now

Reply:
Nice, I'll make some changes to ensure that